File: greedy-algorithms/javascript/set-covering.js
var statesNeeded = new Set(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"]);

const stations = {};
stations["kone"] = new Set(["id", "nv", "ut"]);
stations["ktwo"] = new Set(["wa", "id", "mt"]);
stations["kthree"] = new Set(["or", "nv", "ca"]);
stations["kfour"] = new Set(["nv", "ut"]);
stations["kfive"] = new Set(["ca", "az"]);

const finalStations = new Set();

while (true) {
  let bestStation = null;
  let statesCovered = new Set();

  if (statesNeeded.size) {
    for (let estacao in stations) {
      let states = stations[estacao];
      let covered = new Set([...statesNeeded].filter((x) => states.has(x)));
      if (covered.size > statesCovered.size) {
        bestStation = estacao;
        statesCovered = covered;
      }
    }
  } else {
    break;
  }

  statesNeeded = new Set([...statesNeeded].filter((x) => !statesCovered.has(x)));
  finalStations.add(bestStation);
}

console.log(finalStations);    // Set { 'kone', 'ktwo', 'kthree', 'kfive' }

File: greedy-algorithms/python/set-covering.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-

states_needed = set(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"])

stations = {}
stations["kone"] = set(["id", "nv", "ut"])
stations["ktwo"] = set(["wa", "id", "mt"])
stations["kthree"] = set(["or", "nv", "ca"])
stations["kfour"] = set(["nv", "ut"])
stations["kfive"] = set(["ca", "az"])

final_stations = set()

while True:
  best_station = None
  states_covered = set()
  
  if states_needed:
    for station, states in stations.items():
      covered = states_needed & states
      if len(covered) > len(states_covered):
        best_station = station
        states_covered = covered
  else:
    break
    
  states_needed -= states_covered
  final_stations.add(best_station)

print final_stations   # set(['ktwo', 'kthree', 'kone', 'kfive'])

